samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,109 @@
+   * contains the 6 transformed {@link Message}s.
    *
-   * @param flatMapper  the flat mapper function to map one input {@link Message} to zero or more output {@link Message}s
+   * @param flatMapFn  the flatMap function to map one input {@link Message} to zero or more output {@link Message}
    * @param <OM>  the type of the output {@link Message} in the output {@link MessageStream}
-   * @return the output {@link MessageStream} by applying the map function on the input {@link MessageStream}
+   * @return the output {@link MessageStream} after applying the flatMap function on the input {@link MessageStream}
    */
-  public <OM extends Message> MessageStream<OM> flatMap(Function<M, Collection<OM>> flatMapper) {
-    Operator<OM> op = Operators.getStreamOperator(flatMapper);
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  <OM extends Message> MessageStream<OM> flatMap(FlatMapFunction<M, OM> flatMapFn);
 
   /**
-   * Method to apply a filter function on a {@link MessageStream}
+   * Applies the {@link FilterFunction} to this {@link MessageStream} and returns the resulting {@link MessageStream}.
    *
-   * @param filter  the filter function to filter input {@link Message}s from the input {@link MessageStream}
+   * @param filterFn  the filter function to filter {@link Message}s from the input {@link MessageStream}
    * @return the output {@link MessageStream} after applying the filter function on the input {@link MessageStream}
    */
-  public MessageStream<M> filter(Function<M, Boolean> filter) {
-    Operator<M> op = Operators.<M, M>getStreamOperator(t -> new ArrayList<M>() { {
-        if (filter.apply(t)) {
-          this.add(t);
-        }
-      } });
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  MessageStream<M> filter(FilterFunction<M> filterFn);
 
   /**
-   * Method to send an input {@link MessageStream} to an output {@link org.apache.samza.system.SystemStream}, and allows the output {@link MessageStream}
-   * to be consumed by downstream stream operators again.
+   * Allows sending {@link Message}s in this {@link MessageStream} to an output
+   * {@link org.apache.samza.system.SystemStream} using the provided {@link SinkFunction}.
    *
-   * @param sink  the user-defined sink function to send the input {@link Message}s to the external output systems
+   * This is a terminal transform and does not allow chaining further transforms.
+   *
+   * @param sinkFn  the sink function to send {@link Message}s in this stream to output systems
    */
-  public void sink(VoidFunction3<M, MessageCollector, TaskCoordinator> sink) {
-    this.subscribers.add(Operators.getSinkOperator(sink));
-  }
+  void sink(SinkFunction<M> sinkFn);
 
   /**
-   * Method to perform a window function (i.e. a group-by, aggregate function) on a {@link MessageStream}
+   * Windows this {@link MessageStream} according to the provided {@link Window} semantics (e.g. tumbling,
```

Comment:
Makes sense. Maybe "returns the resulting {@link MessageStream} of {@WindowOutput}"?